### PR TITLE
[SPARK-32084][PYTHON][SQL] Expand dictionary functions

### PR DESCRIPTION
**Pull request by zero323**
### What changes were proposed in this pull request?

- [x] Expand dictionary definitions into standalone functions.
- [x] Fix annotations for ordering functions.

### Why are the changes needed?

To simplify further maintenance of docstrings.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Nice, @zero323. Thanks for working on this.

**Comment by HyukjinKwon** [comment]
> Add proper NumPy-style docstrings to expanded functions.

Oh, let's don't do this in this PR. It should add some dependencies into GitHub Actions, fix the script about `numpydoc`, etc. We could add a dedicated PR for numpydoc only later. Actually I am working (and thinking) on making an initial PR to start this numpydoc style.

**Comment by zero323** [comment]
> > Add proper NumPy-style docstrings to expanded functions.
> 
> Oh, let's don't do this in this PR.

OK. Shall we extend existing docstrings here at all?

And what about the remaining points? Right now functions are in groups, some are not, some ordered by name, some not. Since we introduce quite a few `defs` here, I thought it might be a good idea to sort this stuff.

**Comment by HyukjinKwon** [comment]
Yeah, I think we can just keep docstrings as-is. For grouping stuff, I think it's okay to don't change since it's mainly for code readers. There were similar changes proposed (and abandoned) before, e.g.) https://github.com/apache/spark/pull/27976.

**Comment by HyukjinKwon** [comment]
The current changes look pretty good to go. Let me know when you think it's ready. I'll take one more look and merge it in.

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -42,154 +42,455 @@
...
+
+
+def _invoke_binary_math_function(name, col1, col2):
+    """
+    Invokes binary JVM math function identified by name
+    and wraps the result with :class:`Column`.
+    """
+    return _invoke_function(
+        name,
+        float(col1) if isinstance(col1, numbers.Real) else _to_java_column(col1),
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:77]: I think you can keep the comment about this - converting floats due to the legacy behavior reason.
Also, I think we should keep the previous checking as is. For example, now it won't work with decimals.

```python
>>> isinstance(decimal.Decimal(1), numbers.Real)
False
>>> float(decimal.Decimal(1))
1.0
```

Comment (zero323) [reply python/pyspark/sql/functions.py]: Good catch, thank you!

**Comment by zero323** [comment]
> Yeah, I think we can just keep docstrings as-is. For grouping stuff, I think it's okay to don't change since it's mainly for code readers.

That's true, though I'd argue that it matters, given the size of the module (it is large enough, so `log` defined twice, once through dict and once through def, went unnoticed for years) Not to mention it was clearly the original intention, things just diverged a bit over time.

> There were similar changes proposed (and abandoned) before, e.g.) #27976.

Fair enough. Let's not complicate things here, but I'll allow myself another PR later to see if I can make the case :)

**Comment by zero323** [comment]
Jenkins, retest this please.

**Review by HyukjinKwon — APPROVED** [review]
LGTM. For `asc*` and `desc*`, I believe we can just do in a separate ticket since this PR is already ready to go

**Comment by HyukjinKwon** [comment]
Merged to master.

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -42,154 +41,457 @@
...
+    """
+    return col(col)
+
+
+@since(1.3)
+def asc(col):
+    """
+    Returns a sort expression based on the ascending order of the given column name.
+    """
+    return _invoke_function("asc", col)
```
[the remark this answers is not in the document]
Comment (zero323) [reply python/pyspark/sql/functions.py:119]: I wonder if it makes more sense, for the sake of consistency ,to add `Column` signatures on Scala side.

**Comment by zero323** [comment]
> Merged to master.

Thanks!

File: python/pyspark/sql/functions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zero323) [reply python/pyspark/sql/functions.py:119]: Let's handle this in SPARK-33257

**Pull request closed without merging**